### DISCUSSION TITLE
Support for zyaura CO₂ Sensors

### DISCUSSION THREAD

**Discussion by zivillian** [arendst/Tasmota#14539] [Ideas]
Are there plans, or is anyone working on the support of the Zyaura CO₂ Sensors? The protocol is already [known and documented](https://revspace.nl/CO2MeterHacking) and there is a working [implementation](https://github.com/esphome/esphome/blob/1424091ee51a31d75558f8e4a029ea80edb4d415/esphome/components/zyaura/zyaura.h) for [esphome](https://esphome.io/components/sensor/zyaura.html), but I would like to keep my number of different ESP firmwares as small as possible.

The devices are sold in germany from [TFA Dostmann as AIRCO2NTROL](https://www.tfa-dostmann.de/produkt/co2-monitor-airco2ntrol-mini-31-5006/) but are also available [from other vendors and in other countries](https://www.co2meter.com/collections/desktop/products/co2mini-co2-indoor-air-quality-monitor).

I'd like to port the code from esphome to tasmota, but I've not yet contributed to the project, so I wanted to ask before starting to code.

**Comment by fireball79** [comment]
I need this too, or in easyesp 😁

**Reply by sfromis** [reply]
There is no planned timeline. Support for new sensors depends on when someone submits a PR, which is often someone with interest, having the hardware, and a bit of time to create a driver. Porting from ESPHome is an useful shortcut, of course, but you still have to have the driver match the Tasmota structure for integration, and supplying sensor data for the web UI and JSON payloads.

**Comment by barbudor** [comment]
Looking at the sample arduino code is typical of a code that works fine when running on it's own but which becomes difficult to integrate properly in Tasmota
It is all based on polling the signals in the loop() to process the pseudo-SPI protocol. The ESP is doing only passive listening and has no ability to trigger a message from the sensor at will. 
All fine when you are on your own, a little bit problematic when you have to cooperate with other tasks running on a complex system like Tasmota

There would be a few different strategies:
1. Non blocking polling by polling the signals in the loop whenever Tasmota calls the XdrvNN() function. That's the easiest but make the driver very dependant on the effective "load" of Tasmota. If another task take a little more time like a slow NTP synchronisation, you will loose bits.
2. Blocking polling to read all the 40 bits in a raw. So be the bad guy that will block the other tasks to insure you don't miss a bit. Likely how the DHT type of driver are working except that they can trigger the incoming message. OK if the message duration is short and not to often in order to limit impact on Tasmota. Still a problem to detect the stat of the message in time.
3. Using falling edge interrupt to sample data line and run the FSM in the interrupt. Processing the message in background loop. This need careful implementation as if I'm correct interrupt code must be located in internal ram for cache and performance reasons.
4. Could the hardware SPI and DMA be used? No idea so far


Do you have any info on the clock speed and how often a message is sent by the sensor?

**Comment by fireball79** [comment]
Hi,

i found only this page:
https://www.hackerspace-bamberg.de/Co2_Monitor
This guys decoded the data and there some examples, but i don´t know if you can see the clockspeed.

Maybe here: 
https://hackaday.io/project/5301-reverse-engineering-a-low-cost-usb-co-monitor
This guy decoded it two and then you can see many project based on this in the comments.

I used currently the Co2 Sensor via USB on my rasperrypi and can read this data into FHEM. But i want to use the CO2Meter anywhere in the house.

Some other links:
https://www.co2meter.com/products/co2mini-co2-indoor-air-quality-monitor
https://github.com/schinken/esp8266-co2monitor
https://github.com/schinken/esp8266-co2monitor/blob/de290a1c832ec790b381510863bcc893c4d5df6f/esp8266-co2monitor.ino

greets

**Comment by Jason2866** [comment]
My 2 cents. There is a low chance it gets supported from the active contributors of Tasmota.
The actual known working drivers are blocking. A no go for use with Tasmota. So work has to be done from scratch. There are already Tasmota supported good CO2 sensors. Which are cheaper and easier to source.

**Reply by fireball79** [reply]
Yes, your are right, but the co2 meters I found have only less WAF factor and no displays... So my hope was to use my existing ones, add only a wemos d1 and finish...

How can I ask the active contributors?
My hope was, because there are so many other projects which use this sensor... That porting into tasmota is very easy.

**Reply by sfromis** [reply]
"Asking" active contributors is what you're doing by posting here. And you got a good answer explaining that it may be problematic to port an existing driver. Anyway, a lot of this is up to who is sitting with a strong wish to do it.

Easy to "hide" a good CO2 sensor like SCD41 along with a Wemos D1 mini in a nice wall box. Display may a bit more work, depending on interior design requirements about styling.

**Comment by machinekoder** [comment]
@fireball79 Alternatively connect them to a RPi or similar Linux device: https://github.com/machinekoder/co2watcher

Of course there is more error potential (whole OS that needs to run) and no community support with this solution.

**Comment by fireball79** [comment]
Thanks, but i found another solution. Esphome is supporting this Sensors. Now a wemos d1 with esphome is sending the data via mqtt...

Thats what I want..

So I dont need a usb connection anymore.

VG René

**Comment by machinekoder** [comment]
@fireball79 I think you are referring to a setup like this one, correct? https://foosel.net/blog/2022-01-03-tfa-dostmann-meets-esphome/

Probably wouldn't be too hard to add this feature to Tasmota as well: https://github.com/esphome/esphome/blob/b20bae23ccac3903239ca68bfb3a4add25b7b44d/esphome/components/zyaura/zyaura.cpp

**Reply by fireball79** [reply]
No Problem, meanwhile I use Esphome too, to send values to FHEM via mqtt.
Thanks for answer.
